File: python/tvm/meta_schedule/space_generator/space_generator.py
# Licensed to the Apache Software Foundation (ASF) under one
# or more contributor license agreements.  See the NOTICE file
# distributed with this work for additional information
# regarding copyright ownership.  The ASF licenses this file
# to you under the Apache License, Version 2.0 (the
# "License"); you may not use this file except in compliance
# with the License.  You may obtain a copy of the License at
#
#   http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing,
# software distributed under the License is distributed on an
# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
# KIND, either express or implied.  See the License for the
# specific language governing permissions and limitations
# under the License.
"""
Meta Schedule design space generators that generates design
space for generation of measure candidates.
"""

from collections.abc import Callable, Mapping, Sequence
from typing import TYPE_CHECKING, Literal, TypeAlias

from tvm._ffi import register_object
from tvm.ir import IRModule
from tvm.runtime import Object
from tvm.tir.schedule import Schedule

from .. import _ffi_api

if TYPE_CHECKING:
    from ..mutator import Mutator
    from ..postproc import Postproc
    from ..schedule_rule import ScheduleRule
    from ..tune_context import TuneContext


@register_object("meta_schedule.SpaceGenerator")
class SpaceGenerator(Object):
    """The abstract design space generator interface."""

    ScheduleFnType = (
        Callable[[Schedule], None]  # No output
        | Callable[[Schedule], Schedule]  # Single output
        | Callable[[Schedule], list[Schedule]]  # Multiple outputs
    )

    SpaceGeneratorType: TypeAlias = (
        ScheduleFnType | Literal["post-order-apply", "union"] | "SpaceGenerator"
    )

    postprocs: list["Postproc"] | None
    mutator_probs: dict["Mutator", float] | None

    def _initialize_with_tune_context(self, context: "TuneContext") -> None:
        """Initialize the design space generator with tuning context.

        Parameters
        ----------
        context : TuneContext
            The tuning context for initializing the design space generator.
        """
        _ffi_api.SpaceGeneratorInitializeWithTuneContext(  # type: ignore # pylint: disable=no-member
            self, context
        )

    def generate_design_space(self, mod: IRModule) -> list[Schedule]:
        """Generate design spaces given a module.

        Parameters
        ----------
        mod : IRModule
            The module used for design space generation.

        Returns
        -------
        design_spaces : list[tvm.tir.Schedule]
            The generated design spaces, i.e., schedules.
        """
        return _ffi_api.SpaceGeneratorGenerateDesignSpace(self, mod)  # type: ignore # pylint: disable=no-member

    def clone(self) -> "SpaceGenerator":
        """Clone the design space generator.

        Returns
        -------
        cloned_sg : SpaceGenerator
            The cloned design space generator.
        """
        return _ffi_api.SpaceGeneratorClone(self)  # type: ignore # pylint: disable=no-member

    @staticmethod
    def create(  # pylint: disable=keyword-arg-before-vararg
        kind: Literal["post-order-apply", "union"] | ScheduleFnType = "post-order-apply",
        *args,
        **kwargs,
    ) -> "SpaceGenerator":
        """Create a design space generator."""
        from . import (  # pylint: disable=import-outside-toplevel
            PostOrderApply,
            ScheduleFn,
            SpaceGeneratorUnion,
        )

        if callable(kind):

            def create_schedule_fn(
                func,
                postprocs=[],
                mutator_probs={},
            ):  # pylint: disable=dangerous-default-value
                return ScheduleFn(func, postprocs, mutator_probs)

            return create_schedule_fn(kind, *args, **kwargs)  # type: ignore
        if kind == "post-order-apply":
            return PostOrderApply(*args, **kwargs)
        if kind == "union":
            return SpaceGeneratorUnion(*args, **kwargs)
        if isinstance(kind, str):
            return PostOrderApply(sch_rules=kind, postprocs=kind, mutator_probs=kind)
        raise ValueError(f"Unknown SpaceGenerator: {kind}")


ScheduleFnType = SpaceGenerator.ScheduleFnType
ScheduleRuleType = (
    Sequence["ScheduleRule"] | Literal["llvm", "cuda", "cuda-tensorcore", "hexagon", "from-target"]
)
PostprocType = (
    Sequence["Postproc"] | Literal["llvm", "cuda", "cuda-tensorcore", "hexagon", "from-target"]
)
MutatorProbType = (
    Mapping["Mutator", float] | Literal["llvm", "cuda", "cuda-tensorcore", "hexagon", "from-target"]
)
create = SpaceGenerator.create  # pylint: disable=invalid-name


def _normalize_rules(
    postprocs: PostprocType,
    mutator_probs: MutatorProbType,
) -> tuple[
    Sequence["Postproc"] | None,
    Mapping["Mutator", float] | None,
]:
    # pylint: disable=import-outside-toplevel
    from ..mutator import Mutator
    from ..postproc import Postproc

    # pylint: enable=import-outside-toplevel
    assert postprocs is not None
    assert mutator_probs is not None

    def create_or_none(arg, cls):
        if not isinstance(arg, str):
            return arg
        if arg == "from-target":
            return None
        return cls.create(arg)

    postprocs_ = create_or_none(postprocs, Postproc)
    mutator_probs_ = create_or_none(mutator_probs, Mutator)
    return postprocs_, mutator_probs_


@register_object("meta_schedule.PySpaceGenerator")
class _PySpaceGenerator(SpaceGenerator):
    """
    A TVM object space generator to support customization on the python side.
    This is NOT the user facing class for function overloading inheritance.

    See also: PySpaceGenerator
    """

    def __init__(
        self,
        postprocs: PostprocType = "from-target",
        mutator_probs: MutatorProbType = "from-target",
        f_initialize_with_tune_context: Callable | None = None,
        f_generate_design_space: Callable | None = None,
        f_clone: Callable | None = None,
    ):
        """Constructor."""
        postprocs_, mutator_probs_ = _normalize_rules(postprocs, mutator_probs)

        self.__init_handle_by_constructor__(
            _ffi_api.SpaceGeneratorPySpaceGenerator,  # type: ignore # pylint: disable=no-member
            postprocs_,
            mutator_probs_,
            f_initialize_with_tune_context,
            f_generate_design_space,
            f_clone,
        )


class PySpaceGenerator:
    """
    An abstract space generator with customized methods on the python-side.
    This is the user facing class for function overloading inheritance.

    Note: @derived_object is required for proper usage of any inherited class.
    """

    _tvm_metadata = {
        "cls": _PySpaceGenerator,
        "fields": ["postprocs", "mutator_probs"],
        "methods": ["_initialize_with_tune_context", "generate_design_space", "clone"],
    }

    def _initialize_with_tune_context(self, context: "TuneContext") -> None:
        """Initialize the design space generator with tuning context.

        Parameters
        ----------
        context : TuneContext
            The tuning context for initializing the design space generator.
        """
        raise NotImplementedError

    def generate_design_space(self, mod: IRModule) -> list[Schedule]:
        """Generate design spaces given a module.

        Parameters
        ----------
        mod : IRModule
            The module used for design space generation.

        Returns
        -------
        design_spaces : list[tvm.tir.Schedule]
            The generated design spaces, i.e., schedules.
        """
        raise NotImplementedError

    def clone(self) -> SpaceGenerator:
        """Clone the design space generator.

        Returns
        -------
        cloned_sg : SpaceGenerator
            The cloned design space generator.
        """
        raise NotImplementedError

File: src/meta_schedule/space_generator/space_generator.cc
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
#include "../../target/parsers/aprofile.h"
#include "../utils.h"

namespace tvm {
namespace meta_schedule {

String GetRuleKindFromTarget(const Target& target) {
  if (target->kind->name == "llvm") {
    static const PackedFunc* target_has_feature_fn_ptr =
        runtime::Registry::Get("target.target_has_feature");
    ICHECK(target_has_feature_fn_ptr != nullptr)
        << "The `target.target_has_feature` func is not in tvm registry.";
    bool have_avx512vnni = (*target_has_feature_fn_ptr)("avx512vnni", target);
    bool have_avxvnni = (*target_has_feature_fn_ptr)("avxvnni", target);
    if (have_avx512vnni || have_avxvnni) {
      return "vnni";
    } else {
      bool have_avx512f = (*target_has_feature_fn_ptr)("avx512f", target);
      bool have_avx512bw = (*target_has_feature_fn_ptr)("avx512bw", target);
      if (have_avx512bw && have_avx512f) {
        return "avx512";
      }
    }

    TargetJSON target_json = target::parsers::aprofile::ParseTarget(target->Export());
    TargetFeatures afeatures = Downcast<TargetFeatures>(target_json.at("features"));

    if (Downcast<Bool>(afeatures.at("has_dotprod"))) {
      return "dotprod";
    }
    if (Downcast<Bool>(afeatures.at("has_asimd"))) {
      return "asimd";
    }
    return "llvm";
  }
  if (target->kind->name == "hexagon") {
    return "hexagon";
  }
  if (target->kind->name == "cuda") {
    if (Optional<String> opt_sm = target->GetAttr<String>("arch")) {
      std::string sm = opt_sm.value();
      if (support::StartsWith(sm, "sm_")) {
        sm = sm.substr(3);
        try {
          if (std::stoi(sm) >= 75) {
            return "cuda-tensorcore";
          }
        } catch (const std::invalid_argument& e) {
          LOG(WARNING) << "ValueError: Unable to parse `target.arch`: " << sm
                       << ". Details: " << e.what();
        }
      }
    }
    return "cuda";
  }

  if (IsGPUTarget(target->kind->name)) {
    return "cuda";
  }

  if (target->kind->name == "c") {
    return "c";
  }
  LOG(FATAL) << "Unsupported target: " << target;
  throw;
}

void CloneRules(const SpaceGeneratorNode* src, SpaceGeneratorNode* dst) {
  if (src->postprocs.defined()) {
    Array<Postproc> original = src->postprocs.value();
    Array<Postproc> postprocs;
    postprocs.reserve(original.size());
    for (const Postproc& postproc : original) {
      postprocs.push_back(postproc->Clone());
    }
    dst->postprocs = std::move(postprocs);
  }
  if (src->mutator_probs.defined()) {
    Map<Mutator, FloatImm> original = src->mutator_probs.value();
    Map<Mutator, FloatImm> mutator_probs;
    for (const auto& kv : original) {
      mutator_probs.Set(kv.first->Clone(), kv.second);
    }
    dst->mutator_probs = std::move(mutator_probs);
  }
}

void SpaceGeneratorNode::InitializeWithTuneContext(const TuneContext& context) {
  if (context->target.defined()) {
    String kind = GetRuleKindFromTarget(context->target.value());
    if (!postprocs.defined()) {
      if (kind == "llvm") {
        // for target with skylake-avx512
        postprocs = Postproc::DefaultLLVM();
      } else if (kind == "cuda") {
        postprocs = Postproc::DefaultCUDA();
      } else if (kind == "cuda-tensorcore") {
        postprocs = Postproc::DefaultCUDATensorCore();
      } else if (kind == "hexagon") {
        postprocs = Postproc::DefaultHexagon();
      } else if (kind == "vnni" || kind == "avx512" || kind == "asimd" || kind == "dotprod") {
        postprocs = Postproc::DefaultCPUTensorization();
      } else if (kind == "c") {
        postprocs = Postproc::DefaultMicro();
      } else {
        LOG(FATAL) << "Unsupported kind: " << kind;
        throw;
      }
    }
    if (!mutator_probs.defined()) {
      if (kind == "llvm" || kind == "vnni" || kind == "avx512" || kind == "asimd" ||
          kind == "dotprod") {
        mutator_probs = Mutator::DefaultLLVM();
      } else if (kind == "cuda") {
        mutator_probs = Mutator::DefaultCUDA();
      } else if (kind == "cuda-tensorcore") {
        mutator_probs = Mutator::DefaultCUDATensorCore();
      } else if (kind == "hexagon") {
        mutator_probs = Mutator::DefaultHexagon();
      } else if (kind == "c") {
        mutator_probs = Mutator::DefaultMicro();
      } else {
        LOG(FATAL) << "Unsupported kind: " << kind;
        throw;
      }
    }
  }
  if (postprocs.defined()) {
    for (Postproc i : postprocs.value()) {
      i->InitializeWithTuneContext(context);
    }
  }
  if (mutator_probs.defined()) {
    for (const auto& [mutator, _] : mutator_probs.value()) {
      mutator->InitializeWithTuneContext(context);
    }
  }
}

void PySpaceGeneratorNode::InitializeWithTuneContext(const TuneContext& context) {
  ICHECK(f_initialize_with_tune_context != nullptr)
      << "PySpaceGenerator's InitializeWithTuneContext method not implemented!";
  f_initialize_with_tune_context(context);
}

Array<tir::Schedule> PySpaceGeneratorNode::GenerateDesignSpace(const IRModule& mod) {
  ICHECK(f_generate_design_space != nullptr)
      << "PySpaceGenerator's GenerateDesignSpace method not implemented!";
  return f_generate_design_space(mod);
}

SpaceGenerator PySpaceGeneratorNode::Clone() const {
  ICHECK(f_clone != nullptr) << "PySpaceGenerator's Clone method not implemented!";
  return f_clone();
}

SpaceGenerator SpaceGenerator::PySpaceGenerator(
    Optional<Array<Postproc>> postprocs, Optional<Map<Mutator, FloatImm>> mutator_probs,
    FInitializeWithTuneContext f_initialize_with_tune_context,
    FGenerateDesignSpace f_generate_design_space, FClone f_clone) {
  ObjectPtr<PySpaceGeneratorNode> n = make_object<PySpaceGeneratorNode>();
  n->postprocs = postprocs;
  n->mutator_probs = mutator_probs;
  n->f_initialize_with_tune_context = std::move(f_initialize_with_tune_context);
  n->f_generate_design_space = std::move(f_generate_design_space);
  n->f_clone = std::move(f_clone);
  return SpaceGenerator(n);
}

TVM_REGISTER_OBJECT_TYPE(SpaceGeneratorNode);
TVM_REGISTER_NODE_TYPE(PySpaceGeneratorNode);

TVM_REGISTER_GLOBAL("meta_schedule.SpaceGeneratorInitializeWithTuneContext")
    .set_body_method<SpaceGenerator>(&SpaceGeneratorNode::InitializeWithTuneContext);
TVM_REGISTER_GLOBAL("meta_schedule.SpaceGeneratorGenerateDesignSpace")
    .set_body_method<SpaceGenerator>(&SpaceGeneratorNode::GenerateDesignSpace);
TVM_REGISTER_GLOBAL("meta_schedule.SpaceGeneratorPySpaceGenerator")
    .set_body_typed(SpaceGenerator::PySpaceGenerator);
TVM_REGISTER_GLOBAL("meta_schedule.SpaceGeneratorClone")
    .set_body_method<SpaceGenerator>(&SpaceGeneratorNode::Clone);

}  // namespace meta_schedule
}  // namespace tvm
